# heart_failure_analysis
For this project, we aim to analyse publicly available gene expression datasets from patient samples to understand disease mechanisms and identify key regulatory genes and pathways using various bioinformatic tools and integration with ChIP-Seq/CLIP-Seq data.


1. Working with publicly available RNA-seq data from the NCBI GEO database:
Explore the GEO series and describe the experimental design and the appropriate controls
and disease samples to be compared. Refer to the appropriate publications/ literature (if
available) to understand the background of the study, how the data was generated, and
the processing pipeline and annotation used.
You do not need to run analysis of all the samples. Subset the analysis to a specific tissue-
type of interest and/or disease subtype and/or other relevant conditions. The key task here
is to choose an appropriate comparison to make as per experimental design and to
formulate your research question as a group. Each team member to pick a different subset.
2. Perform exploratory data analysis.
Run exploratory data analysis (EDA) to identify if the samples are appropriate for further
analysis. Check the data transformations, normalizations, batch effects, and variance
distribution and apply appropriate corrections if relevant. Explain the results of EDA.
3. Perform differential expression analysis using appropriate tools:
Run Differential Gene Expression Analysis (DGEA) using the appropriate model, test and
tools with reasoning for your choice. Generate the appropriate diagnostic plots like volcano
plots and explain what these diagnostic plots imply about your data and analysis. How
many significantly differentially expressing genes (DEGs) does your model output? Does
the profile change on running differential transcript utilization or differential transcript
expression analysis? Why? Which method fits your research question best?
4. Gain biological insights using downstream functional analysis:
Run the downstream functional analysis such as overrepresentation, enrichment, leading
edge, transcription factor analysis using the appropriate gene sets and tools to answer
your research question. Identify key pathways and regulatory genes affecting the disease
pathology. Cross-validate your findings with literature.
5. Integration with ChIP-Seq/CLIP-Seq data
Pick a transcription factor or RNA binding protein of interest to your disease subtype from
literature. This will be the same protein for all members in a group. Find the relevant ChIP-
Seq or CLIP-Seq data from ENCODE or POSTAR3 respectively. Are the targets of this
protein dysregulated in your dataset? What conclusions can you make regarding the role
of the protein towards disease pathogenesis? Provide relevant visualization(s).
6. Meta-analysis of subtypes chosen by group members â€“ for presentation only.
How similar or different are the results from the subtypes chosen to be analysed by each
group member? Are there common mechanisms across the subtypes?
